### PR TITLE
FEAT: Support processing files in parallel

### LINKED ISSUE #3: Parallelize flag to increase performance when multiple files provided
Hi, thank you for your great work! This has been quite useful to me. I'm incorporating as part of a git hook and ideally would like the hook to be as fast as possible.

I haven't taken a close look at the code yet but I don't think there's an option for running the formating in parallel for every file provided.

I think something like `--parallel {n}` where n is the number of parallel files to do at a time, where if either a special character is provided or no value is provided at all, the maximum/all the files are processed in parallel.

What are your thoughts on this? And may I ask if you're happy for margin a PR?

### PR DESCRIPTION
**Pull request by amirha97**
A rough outline of the feature. 

I did some locally testing with 46 files: without the flag, it takes 7-8 seconds, with the `parallel` flag it takes only 2 seconds. 

I'll investigate a bit more how good this approach is and if there are better ways of doing it and generally improving performance

Fixes #3

### CODE REVIEW

**Comment by nkmathew** [comment]
It's impressively fast. Tried it on 192 files from Norvig's [site](http://norvig.com/ltd/test/) and the Common Lisp Cookbook - 21 vs 3 mins.

+ Mind adding a docstring to the wrapper to please `pylint`?
+ I'm noticing the parameter isn't taking arguments for the number of worker processes like you specified in the issue. Have you decided against it?

**Comment by amirha97** [comment]
> It's impressively fast. Tried it on 192 files from Norvig's [site](http://norvig.com/ltd/test/) and the Common Lisp Cookbook - 21 vs 3 mins.
> 
> * Mind adding a docstring to the wrapper to please `pylint`?
> * I'm noticing the parameter isn't taking arguments for the number of worker processes like you specified in the issue. Have you decided against it?

Good to hear! 
I definitely need to add documentation. 
I've been doing some reading and think multiprocessing is actually not the best way to go about it. Using ayncio should make it even faster. The only issue is that it will require newer versions of python I believe is this okay with you?

Regarding adding an option to specify the number of concurrent processes, I thought that can be added in the future if necessary. If we go the asyncio way, I'm not sure we can easily add support for restricting that and there would be no benefit in doing that I believe, as the operation is io-bound.

I also think it would be good if I can add some kind of test to measure performance for any future work and uses, but not sure how easy this will be todo. I think I'll take a subset of the Norvig site files you linked as realistic data, (haven't yet looked at the files or checked the license) Do you have any suggestions regarding this?

TODOs:
- [ ] Add documentation
- [ ] Use Async io instead of multiprocessing and measure performance
- [ ] Add a way of testing performance automatically

**Comment by nkmathew** [comment]
Python 2 reached EOL on new year's so I guess it's safe to move on. And if you've
moved on to Python 3 then you'd most likely be using a version that is not 5+ years
old (v3.4)

I also did some reading on AsyncIO and I'm skeptical it'll beat the speed
multiprocessing gives but I'm happy to be proven wrong. From what I've read asyncio
would thrive where the io block is significant enough. I'm not sure reading and
writing text files that rarely exceed 100kb comes close. The excerpts I've seen on
discussions on this mention network IO being a good use case.

The scary part is the implementation. I looked into how black does [it][1] and on
first glance it does not look like a fun exercise. Still I'm really curious to see
how this will turn out. It sounds like a good excuse to use asyncio. We have
multiprocessing as a fallback after all.

The files on Norvig website seem to be still under this [license][2] though it says
*1998-2002*. Think it's a fair assumption to say that it still applies

I'm also not sure how you'd do a performance comparison test. Can't say I've ever
done it. Testing looks even harder with asynchronous code. I'm drawing a blank here,
sorry

[1]: https://github.com/psf/black/blob/c7da3482c7badf5c6d8c0c9a37495ab4529eaaad/src/black/__init__.py
[2]: https://norvig.com/license.html

File: yasi.py [hunk 1]
```diff
@@ -1169,7 +1183,7 @@ def _after_indentation(res, options=None, fpath=''):
     indented_code = res['indented_code']
     indent_result = ''.join(indented_code)
     if indented_code == res['original_code'] and opts.files:
-        message = "\nFile `%s' has already been formatted. Leaving it unchanged. . .\n"
+        message = "File '%s' has already been formatted. Leaving it unchanged. . .\n"
```
Comment (amirha97) [inline yasi.py:1186]: This is more of personal preference, the output is a bit harder to read with 2 line endings. Let me know if you're not a fan

**Pull request closed without merging**